packages/classes/src/PayloadArray.js

```diff
@@ -0,0 +1,98 @@
+import Payload from './Payload'
+
+/**
+ * PayloadArray stores a series of Payloads in a data structure that allows for easy filtering
+ * and lookup by position and type.
+ */
+class PayloadArray {
+	payloads = []
+
+	constructor(...payloads) {
+		[...payloads].forEach((payload) => this.insert(payload))
+	}
+
+	get length() { return this.payloads.length }
+
+	/**
+	 * Insert a Payload into the PayloadArray while maintaining position order.
+	 *
+	 * @param  {Payload} payload The Payload to be inserted.
+	 * @return {PayloadArray}    This PayloadArray to allow for chaining.
+	 */
+	insert = (payload) => {
+		if (!(payload instanceof Payload)) {
+			throw new Error('Attempt to insert a non-Payload to PayloadArray')
+		}
+		const index = this.indexOfPosition(payload.position)
+		this.payloads.splice(index, 0, payload)
+		return this
+	}
+
+	/**
+	 * Find the index of the first Payload whose position is greater than or equal to this one.
+	 *
+	 * NOTE: if no qualifying payloads exist, this will return the length of the PayloadArray
+	 *
+	 * @param  {Number} position The position being searched for.
+	 * @return {Number}          The index of the item, or the PayloadArray length.
+	 */
+	indexOfPosition = (position) => {
+		const recursiveSearch = (sortedPayloads, target, left, right) => {
+			if (left === right) return left
+			const index = Math.floor((left + right) / 2)
+			if (index === sortedPayloads.length) return index
+			const payload = sortedPayloads[index]
+			if (payload.position === target) return index
+			if (payload.position < target) {
+				return recursiveSearch(sortedPayloads, target, index + 1, right)
+			}
+			if (payload.position > target) return recursiveSearch(sortedPayloads, target, left, index)
+			return index
+		}
+		return recursiveSearch(this.payloads, position, 0, this.payloads.length)
+	}
+
+	/**
+	 * Filter out the Payloads of the specified type.
+	 *
+	 * NOTE: This returns a new PayloadArray and does not modify the original.
+	 *
+	 * @param  {String} type  The type of payload to include in the new PayloadArray.
+	 * @return {PayloadArray} A filtered PayloadArray.
+	 */
+	filterByType = (type) => {
+		const filteredPayloadArray = new PayloadArray()
+		filteredPayloadArray.payloads = this.payloads.filter(
+			(payload) => RegExp(`^${type.replace(/\.ANY$/, '.*')}$`).test(payload.type),
+		)
+		return filteredPayloadArray
+	}
+
+	/**
+	 * Filter out the Payloads that exist within a specified duration.
+	 *
+	 * NOTE: This returns a new PayloadArray and does not modify the original.
+	 *
+	 * @param  {Number} start The start position.
+	 * @param  {Number} end   The end position.
+	 * @return {PayloadArray} A filtered PayloadArray.
+	 */
+	filterByPosition = (start, end = null) => {
+		const filteredPayloadArray = new PayloadArray()
+		filteredPayloadArray.payloads = this.payloads.filter(
```

Comment:
There's a possible optimization here if we're confident the PayloadArray will always be sorted by position, using a combination of `findIndex()` and `slice()`.

Finding the last element that matches `end` is a tiny bit trickier since there's no native `findLastIndex()`, but both [Underscore](https://underscorejs.org/#findLastIndex)/[Lodash](https://lodash.com/docs/4.17.15#findLastIndex) have examples we could import or replicate. (This could also make `indexOfPosition()` faster and more legible.)

I think that falls squarely into optimization territory and not requirement, so… just noting it for now! But that would get away from having to necessarily examine every single array element the way `filter()` does.

Reply:
One note as I think through what you're suggesting: it's not necessarily the case that a payload exists with `end` -- payloads span time, so end might fall inside of a payload's duration but not actually be represented as the `position` of that payload

Reply:
That's true, but (so long as the payload array is always sorted) we can know that with a quick test of the last element before actually looking for the "end", right?

Reply:
I'm not sure!

So say we have this list of payloads (writing in terms of position spans:

- { position: 0, duration: 10 }
- { position: 11, duration: 4 }
- { position: 16, duration: 10}

`filterByPosition(1,12)` should return the first two Payloads -- even though neither `1` nor `12` actually exist as positions.

We should also remember that it isn't necessarily the case that the payloads completely fill the time (there might be gaps of duration that have no payload at all)

Reply:
Ah okay, so, that basically means finding the start means looking for `payload.position >= start` while finding the end is `(payload.position + payload.duration) <= end`, right?

Reply:
Almost -- take out the duration in the end calculation.

And indeed that's what is reflected:

```
payload.position >= start
&& (end === null
|| payload.position <= end)
```

Reply:
(My suggestion is more about getting away from `filter()`, which necessarily looks at all array elements, than changing how you're actually identifying start/end points. But since you brought it up, it _does_ look like your endpoint-finder needs to account for duration!)

Reply:
I'm so confused!

We're having two conversations, right? Identifying the endpoint, and optimally filtering the array. My original suggestion was about optimally filtering the array, but set that aside for a second.

On identifying the endpoint: doesn't `payload.position` note the _start_ of the payload's timestamp? According to your example (where `{ position: 11, duration: 4 }` should be included in `filterByPosition(1,12)`, wouldn't `payload.position <= end` be insufficient?

Reply:
OH MY GOD

Reply:
I'm unbelievably foggy-headed today, evidently. I'm sorry.

Reply:
You're fine! This is the micro-algorithm kind of stuff that people solved for like 25 years ago and generally we try to avoid because of exactly the journey you just went through!

Point is: you're not foggy, it's just that data structures should generally be borrowed, not built, so we don't have to think about these things.

Reply:
Just to capture an issue you flagged in slack: The code as written would not pass the example I gave.  Specifically, it would NOT include the first payload in my example, because the start time of that payload  (`0`) is BEFORE the passed start (`1`).

We need to decide if we want this to give payloads that exist before the specified start.  My inclination is no, and that the returned payloads should be all payloads whose `position` values exist within the specified range.

Reply:
Okay, that makes sense. I think I did assume that `filterByPosition()` would work like this theoretical `filterByTimeWindow()`, but if the current need is for the behavior you describe, then yeah let's wait until we need it.

For the record of future historians, in Slack I had flagged that a test of the previous [example above](https://github.com/tvkitchen/base/pull/46#discussion_r441842186) actually failed currently:

```js
it('should return a PayloadArray whose payloads exist within the specified time range', () => {
	const payloadArray = new PayloadArray(
		new Payload({ position: 0, duration: 10 }),
		new Payload({ position: 11, duration: 4 }),
		new Payload({ position: 16, duration: 10 }),
	)
	const filteredPayloadArray = payloadArray.filterByPosition(1, 12)
	expect(filteredPayloadArray).toBeInstanceOf(PayloadArray)
	expect(filteredPayloadArray.length).toBe(2)
})
```

But that's okay; we really do want `filterByPosition()` to just provide those payloads whose `position` itself (not entire `position + duration`) are in the provided window.

Reply:
I'll think a bit more about the position vs timewindow -- I think if we *name* it fileterByPosition it would be wierd to return Payloads whose `position` attributes do not fit inside of the time window (in terms of making sense from a literal perspective)

So basically, I'll think about what fits the use case better and implement that one (and if in thinking about that I think of a use case for both, I'll implement both!)

Reply:
So back to my _original_ conversation here: is using the array `filter` method the most performant way to perform this filter? It has to examine every single array item, when we can short-cut that by finding the first payload with a position after/equal to `start`, the last payload with a position before/equal `end`, and using those as our slice boundaries. Right?

It's obvs a performance optimization but to [your point in the `indexOfPosition()` discussion](https://github.com/tvkitchen/base/pull/46#discussion_r441834079), this is a core method where performance optimizations actually matter.

Reply:
Ah yes!  That point is not lost and regardless, I'll aim to make this better optimized (possibly by even leveraging indexOfPosition).

Reply:
This has been re-implemented to use the `indexOfPosition` which I believe is going to be close enough to fastest and good enough for now.